File: 400-499/447/447.cpp
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
private:
	int getDistance(vector<vector<int>>& points, int i, int j) {
		int x = points[i][0] - points[j][0];
		int y = points[i][1] - points[j][1];
		return x * x + y * y;
	}

public:
	int numberOfBoomerangs(vector<vector<int>>& points) {
		int total = 0;
		
		for (int i = 0; i < points.size(); i++) {
			// Cache of counts of distances
			unordered_map<int, int> cache;
			
			// For each point i, iterate all other points
			for (int j = 0; j < points.size(); j++) {
				int dist = getDistance(points, i, j);

				// Assuming the current count of this distance is n, then:
				// - number of boomerangs before current increment = n * (n - 1)
				// - number of boomerangs after current increment  = (n + 1) * n
				// The diff is n * 2, so we just need to add n * 2 to get the new number of
				// boomerangs from the old number
				if (cache[dist] > 0)
					total += cache[dist] * 2;

				// Increment count of this distance
				cache[dist] += 1;
			}
		}
		
		return total;
	}
};

File: 0400-0499/447/447.go
func numberOfBoomerangs(points [][]int) int {
	var total int = 0
	
	for _, i := range points {
		// Cache of counts of distances
		var cache map[int]int = map[int]int {}
		
		// For each point i, iterate all other points
		for _, j := range points {

			// Calculate distance
			x := i[0] - j[0]
			y := i[1] - j[1]
			dist := x * x + y * y
			
			if _, ok := cache[dist]; ok {
				// Assuming the current count of this distance is n, then:
				// - number of boomerangs before current increment = n * (n - 1)
				// - number of boomerangs after current increment  = (n + 1) * n
				// The diff is n * 2, so we just need to add n * 2 to get the new number of
				// boomerangs from the old number
				total += cache[dist] * 2

				// Increment count of this distance
				cache[dist] += 1
			} else {

				// Init count of this distance to 1
				cache[dist] = 1
			}
		}
	}
	
	return total
}
